### PR TITLE
[cli] - IP Allow list command 

### PR DESCRIPTION
**Pull request by Shridhad**
closes: https://github.com/neondatabase/cloud/issues/8882

**New Commands:** 

1. `neonctl ip-allow list --projectId <projectId>`
2. `neonctl ip-allow add 192.168.1.1 127.0.0.1-127.0.0.2 --projectId <projectId>`
3. `neonctl ip-allow add 192.168.1.1 --primary-only false --projectId <projectId>`
4. `neonctl ip-allow remove 192.168.1.1 --projectId <projectId>` 
5. `neonctl ip-allow reset --projectId <projectId>`

ip-allow-command-help

https://github.com/neondatabase/neonctl/assets/1949852/092e74ea-b61d-4215-b8c0-f4d5a2cebecb

### CODE REVIEW

File: src/commands/ip_allow.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  'primary_branch_only',
+] as const;
+
+export const command = 'ip-allow';
+export const describe = 'Manage IP Allow';
+export const builder = (argv: yargs.Argv) => {
+  return argv
+    .usage('$0 ip-allow <sub-command> [options]')
+    .options({
+      'project-id': {
```
Comment (Shridhad) [inline src/commands/ip_allow.ts:21]: @duskpoet @lpetkov Earlier I suggested `projectId` as positional argument [here](https://github.com/neondatabase/cloud/issues/8882#issuecomment-1891930122), but looking at other commands, (other than `projects`) we pass the `projectId` as option. Example - [Branch List](https://neon.tech/docs/reference/cli-branches#list) and [Connection String](https://neon.tech/docs/reference/cli-connection-string#the-connection-string-command)

Keeping it consistent with other commands as Ed suggested.

File: src/commands/projects.ts [hunk 2]
```diff
@@ -85,13 +85,15 @@ export const builder = (argv: yargs.Argv) => {
                 .description,
             type: 'string',
             array: true,
+            deprecated: "Deprecated. Use 'ip-allow' command",
```
Comment (Shridhad) [inline src/commands/projects.ts:88]: Deprecating `--ip-allow` and `--ip-primary-only` options from the `projects update` command for now. We can remove it later in future.

File: src/commands/ip_allow.ts (hunk 1 above)
Comment (danieltprice) [reply src/commands/ip_allow.ts:21]: Hey, I noticed that you have "--projectId", but elsewhere we use "--project-id":

neonctl branches list --project-id square-cell-51292327

**Comment by danieltprice** [comment]
Hey, I noticed that you have "--projectId", but elsewhere we use "--project-id":

neonctl branches list --project-id square-cell-51292327

**Comment by Shridhad** [comment]
> Hey, I noticed that you have "--projectId", but elsewhere we use "--project-id":

@danieltprice Both `projectId` and `project-id` work for all commands (`branches list` as well). The command line tool we use will convert `project-id` to `projectId` in the code.

**Comment by danieltprice** [comment]
I suggest these description changes. Capitalize first letter in the sentence in all cases & we use "allowlist" in the docs:

List the IP allowlist
Add IP addresses to the IP allowlist
Remove IP addresses from the IP allowlist
Reset the IP allowlist

**Review by danieltprice — APPROVED** [review]
Some minor suggestions for the command descriptions.

File: src/commands/ip_allow.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+  writer(props).end(parse(response.project, props), {
+    fields: IP_ALLOW_FIELDS,
+  });
+};
+
+const remove = async (props: ProjectScopeProps & { ips: string[] }) => {
+  if (props.ips.length <= 0) {
+    log.error(
+      `Remove individual IP addresses and ranges. Example: neonctl ip-allow remove 192.168.1.1 --projectId <projectId>`,
```
Comment (danieltprice) [inline src/commands/ip_allow.ts:141]:
```suggestion
      `Remove individual IP addresses and ranges. Example: neonctl ip-allow remove 192.168.1.1 --project-id <id>`,
```

File: src/commands/ip_allow.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+const add = async (
+  props: CommonProps &
+    ProjectScopeProps & {
+      ips: string[];
+      primaryOnly?: boolean;
+    },
+) => {
+  if (props.ips.length <= 0) {
+    log.error(`Enter individual IP addresses, define ranges with a dash, or use CIDR notation for more flexibility.
+       Example: neonctl ip-allow add 192.168.1.1, 192.168.1.20-192.168.1.50, 192.168.1.0/24 --projectId <projectId>`);
```
Comment (danieltprice) [inline src/commands/ip_allow.ts:110]:
```suggestion
       Example: neonctl ip-allow add 192.168.1.1, 192.168.1.20-192.168.1.50, 192.168.1.0/24 --project-id <id>`);
```

File: src/commands/ip_allow.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    project,
+  });
+
+  writer(props).end(parse(data.project, props), {
+    fields: IP_ALLOW_FIELDS,
+  });
+
+  if (props.ips.length <= 0) {
+    log.info(
+      `The IP allow list has been reset. All databases on project "${data.project.name}" are now exposed to the internet`,
```
Comment (danieltprice) [inline src/commands/ip_allow.ts:189]:
```suggestion
      `The IP allowlist has been reset. All databases on project "${data.project.name}" are now exposed to the internet`,
```

File: src/commands/ip_allow.test.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    expected: {
+      snapshot: true,
+    },
+  });
+
+  testCliCommand({
+    name: 'Remove IP allow - Error',
+    args: ['ip-allow', 'remove', '--projectId', 'test'],
+    expected: {
+      stderr: `ERROR: Remove individual IP addresses and ranges. Example: neonctl ip-allow remove 192.168.1.1 --projectId <projectId>
```
Comment (danieltprice) [inline src/commands/ip_allow.test.ts:43]:
```suggestion
      stderr: `ERROR: Remove individual IP addresses and ranges. Example: neonctl ip-allow remove 192.168.1.1 --project-id <id>
```

File: src/commands/ip_allow.test.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    name: 'Reset IP allow',
+    args: ['ip-allow', 'reset', '--projectId', 'test'],
+    expected: {
+      snapshot: true,
+      stdout: `id: test
+name: test_project
+IP_addresses: []
+primary_branch_only: false
+`,
+      stderr: `INFO: The IP allow list has been reset. All databases on project "test_project" are now exposed to the internet
```
Comment (danieltprice) [inline src/commands/ip_allow.test.ts:66]:
```suggestion
      stderr: `INFO: The IP allowlist has been reset. All databases on project "test_project" are now exposed to the internet
```

File: src/commands/ip_allow.ts [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+export const command = 'ip-allow';
+export const describe = 'Manage IP Allow';
+export const builder = (argv: yargs.Argv) => {
+  return argv
+    .usage('$0 ip-allow <sub-command> [options]')
+    .options({
+      'project-id': {
+        describe: 'Project ID',
+        type: 'string',
+        demandOption: true,
```
Comment (duskpoet) [inline src/commands/ip_allow.ts:24]: Don't make it demanded option, this way context would not work.

File: src/commands/ip_allow.ts [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+      },
+    })
+    .command(
+      'list',
+      'List IP allow list',
```
Comment (danieltprice) [inline src/commands/ip_allow.ts:29]:
```suggestion
      'List IP the allowlist',
```

File: src/commands/ip_allow.ts [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      (yargs) => yargs,
+      async (args) => {
+        await list(args as any);
+      },
+    )
+    .command(
+      'add [ips...]',
+      'Add IP addresses to IP allow list',
```
Comment (danieltprice) [inline src/commands/ip_allow.ts:37]:
```suggestion
      'Add IP addresses to the IP allowlist',
```

File: src/commands/ip_allow.ts [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+              type: 'boolean',
+            },
+          }),
+      async (args) => {
+        await add(args as any);
+      },
+    )
+    .command(
+      'remove [ips...]',
+      'remove IP addresses from IP allow list',
```
Comment (danieltprice) [inline src/commands/ip_allow.ts:62]:
```suggestion
      'Remove IP addresses from the IP allowlist',
```

File: src/commands/ip_allow.ts [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+          default: [],
+          array: true,
+        }),
+      async (args) => {
+        await remove(args as any);
+      },
+    )
+    .command(
+      'reset [ips...]',
+      'reset IP allow list',
```
Comment (danieltprice) [inline src/commands/ip_allow.ts:76]:
```suggestion
      'Reset the IP allowlist',
```

File: src/commands/ip_allow.ts [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    );
+  }
+};
+
+const parse = (project: Project, props: CommonProps) => {
+  const ips = project.settings?.allowed_ips?.ips ?? [];
+  return {
+    id: project.id,
+    name: project.name,
+    IP_addresses: props.output === 'table' ? ips.join('\n') : ips,
```
Comment (duskpoet) [inline src/commands/ip_allow.ts:199]: Let's add this to the writer? I'd debate that all array fields should be treated like this

**Review by duskpoet — COMMENTED** [review]
Nice!

File: src/commands/ip_allow.ts (hunk 8 above)
Comment (Shridhad) [reply src/commands/ip_allow.ts]: Good catch. Removed.

File: src/commands/ip_allow.ts (hunk 13 above)
Comment (Shridhad) [reply src/commands/ip_allow.ts]: Moved to writer. Also added support for JSON Object in `table` output format.

**Review by duskpoet — APPROVED** [review]

**Pull request merged**